### PR TITLE
cmake: Reflect minor version change in SONAME for C++ and C#

### LINKED ISSUE #23205: libgrpc++ breaks ABI without changing the soname
https://bugs.debian.org/955279
`
The _ZN4grpc13ClientContextC1Ev symbol for grpc::ClientContext::ClientContext() was included in
libgrpc++1 versions up to 1.17.2-1.
Since version 1.22.0-2 it is missing from that library.
`

Commit d166e38e appears to be a revert of a (partial?) soname bump for libgrpc++, such a soname bump would fix this problem.

### PR DESCRIPTION
**Pull request by weltling**
This helps to localize the ABI changes between versions and makes
possible to install different versions of GRPC side-by-side on the same
root filesystem. This is still not an ideal solution, but at least the
expectation would be whenever the core version has changed, the SONAME
would be different for any other related artifacts.

Closes #23205 and many others as per a query here https://github.com/grpc/grpc/pulls?q=is%3Apr+soname . Now that the version handling is centralized and the CMake infrastructure is used.

Signed-off-by: Anatol Belski <anbelski@linux.microsoft.com>

A more extended explanation. Please see the artifact names from a build using `mkdir build; cd build; cmake -DBUILD_SHARED_LIBS=on ..` without this patch:

```bash
ls -lGgaH | grep '\.so'
lrwxrwxrwx  1       24 Mar  4 16:37 libaddress_sorting.so -> libaddress_sorting.so.15
lrwxrwxrwx  1       28 Mar  4 16:37 libaddress_sorting.so.15 -> libaddress_sorting.so.15.0.0
-rwxrwxr-x  1    18736 Mar  4 16:37 libaddress_sorting.so.15.0.0
lrwxrwxrwx  1       12 Mar  4 16:38 libgpr.so -> libgpr.so.15
lrwxrwxrwx  1       16 Mar  4 16:38 libgpr.so.15 -> libgpr.so.15.0.0
-rwxrwxr-x  1   179192 Mar  4 16:38 libgpr.so.15.0.0
lrwxrwxrwx  1       14 Mar  4 16:38 libgrpc++.so -> libgrpc++.so.1
lrwxrwxrwx  1       23 Mar  4 16:38 libgrpc++.so.1 -> libgrpc++.so.1.37.0-dev
-rwxrwxr-x  1  2789096 Mar  4 16:38 libgrpc++.so.1.37.0-dev
lrwxrwxrwx  1       19 Mar  4 16:38 libgrpc++_alts.so -> libgrpc++_alts.so.1
lrwxrwxrwx  1       28 Mar  4 16:38 libgrpc++_alts.so.1 -> libgrpc++_alts.so.1.37.0-dev
-rwxrwxr-x  1   144864 Mar  4 16:38 libgrpc++_alts.so.1.37.0-dev
lrwxrwxrwx  1       28 Mar  4 16:38 libgrpc++_error_details.so -> libgrpc++_error_details.so.1
lrwxrwxrwx  1       37 Mar  4 16:38 libgrpc++_error_details.so.1 -> libgrpc++_error_details.so.1.37.0-dev
-rwxrwxr-x  1     7416 Mar  4 16:38 libgrpc++_error_details.so.1.37.0-dev
lrwxrwxrwx  1       25 Mar  4 16:38 libgrpc++_reflection.so -> libgrpc++_reflection.so.1
lrwxrwxrwx  1       34 Mar  4 16:38 libgrpc++_reflection.so.1 -> libgrpc++_reflection.so.1.37.0-dev
-rwxrwxr-x  1  1120672 Mar  4 16:38 libgrpc++_reflection.so.1.37.0-dev
lrwxrwxrwx  1       23 Mar  4 16:38 libgrpc++_unsecure.so -> libgrpc++_unsecure.so.1
lrwxrwxrwx  1       32 Mar  4 16:38 libgrpc++_unsecure.so.1 -> libgrpc++_unsecure.so.1.37.0-dev
-rwxrwxr-x  1  2418208 Mar  4 16:38 libgrpc++_unsecure.so.1.37.0-dev
lrwxrwxrwx  1       13 Mar  4 16:38 libgrpc.so -> libgrpc.so.15
lrwxrwxrwx  1       17 Mar  4 16:38 libgrpc.so.15 -> libgrpc.so.15.0.0
-rwxrwxr-x  1 15887920 Mar  4 16:38 libgrpc.so.15.0.0
lrwxrwxrwx  1       24 Mar  4 16:38 libgrpc_csharp_ext.so -> libgrpc_csharp_ext.so.15
lrwxrwxrwx  1       28 Mar  4 16:38 libgrpc_csharp_ext.so.15 -> libgrpc_csharp_ext.so.15.0.0
-rwxrwxr-x  1    52272 Mar  4 16:38 libgrpc_csharp_ext.so.15.0.0
lrwxrwxrwx  1       27 Mar  4 16:38 libgrpc_plugin_support.so -> libgrpc_plugin_support.so.1
lrwxrwxrwx  1       36 Mar  4 16:38 libgrpc_plugin_support.so.1 -> libgrpc_plugin_support.so.1.37.0-dev
-rwxrwxr-x  1   925048 Mar  4 16:38 libgrpc_plugin_support.so.1.37.0-dev
lrwxrwxrwx  1       22 Mar  4 16:38 libgrpc_unsecure.so -> libgrpc_unsecure.so.15
lrwxrwxrwx  1       26 Mar  4 16:38 libgrpc_unsecure.so.15 -> libgrpc_unsecure.so.15.0.0
-rwxrwxr-x  1  8170872 Mar  4 16:38 libgrpc_unsecure.so.15.0.0
lrwxrwxrwx  1       23 Mar  4 16:38 libgrpcpp_channelz.so -> libgrpcpp_channelz.so.1
lrwxrwxrwx  1       32 Mar  4 16:38 libgrpcpp_channelz.so.1 -> libgrpcpp_channelz.so.1.37.0-dev
-rwxrwxr-x  1  1867712 Mar  4 16:38 libgrpcpp_channelz.so.1.37.0-dev
lrwxrwxrwx  1       12 Mar  4 16:37 libupb.so -> libupb.so.15
lrwxrwxrwx  1       16 Mar  4 16:37 libupb.so.15 -> libupb.so.15.0.0
-rwxrwxr-x  1   228936 Mar  4 16:37 libupb.so.15.0.0
$ objdump -p libgrpc++.so.1.37.0-dev | grep SONAME
  SONAME               libgrpc++.so.1
$ readelf -a libgrpc++_reflection.so | grep NEED
  [ 6] .gnu.version_r    VERNEED          00000000000581a8  000581a8
 0x0000000000000001 (NEEDED)             Shared library: [libgrpc++.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libprotobuf.so.3.15.2.0]
 0x0000000000000001 (NEEDED)             Shared library: [libstdc++.so.6]
 0x0000000000000001 (NEEDED)             Shared library: [libgcc_s.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libc.so.6]
 0x000000006ffffffe (VERNEED)            0x581a8
 0x000000006fffffff (VERNEEDNUM)         3
```

Especially as in all the previous releases, the versionless DSO is linked  to `libgrpc++.so -> libgrpc++.so.1`, and that one is linked to  `libgrpc++.so.1 -> libgrpc++.so.1.37.0-dev` . I was looking as far behind as to 1.27.0 which has same SONAME as 1.37.0-dev. That's where the point of issue lays - while any consuming artifact will link with `libgrpc++.so.1`, the actual versions of GRPC are different. This has two issues:

- It is not possible to move a consuming artifact to another system as it is not known which exact version `libgrpc++.so.1` points to
- It is not possible to install different versions of GRPC side-by-side on the same root filesystem

Now please look at the artifacts produced using this patch:

```bash
$ ls -lGgaH | grep '\.so'
lrwxrwxrwx  1       24 Mar  4 16:12 libaddress_sorting.so -> libaddress_sorting.so.15
lrwxrwxrwx  1       28 Mar  4 16:12 libaddress_sorting.so.15 -> libaddress_sorting.so.15.0.0
-rwxrwxr-x  1    18736 Mar  4 16:12 libaddress_sorting.so.15.0.0
lrwxrwxrwx  1       12 Mar  4 16:12 libgpr.so -> libgpr.so.15
lrwxrwxrwx  1       16 Mar  4 16:12 libgpr.so.15 -> libgpr.so.15.0.0
-rwxrwxr-x  1   179192 Mar  4 16:12 libgpr.so.15.0.0
lrwxrwxrwx  1       15 Mar  4 16:13 libgrpc++.so -> libgrpc++.so.15
-rwxrwxr-x  1  2789096 Mar  4 16:13 libgrpc++.so.1.37.0-dev
lrwxrwxrwx  1       23 Mar  4 16:13 libgrpc++.so.15 -> libgrpc++.so.1.37.0-dev
lrwxrwxrwx  1       20 Mar  4 16:13 libgrpc++_alts.so -> libgrpc++_alts.so.15
-rwxrwxr-x  1   144864 Mar  4 16:13 libgrpc++_alts.so.1.37.0-dev
lrwxrwxrwx  1       28 Mar  4 16:13 libgrpc++_alts.so.15 -> libgrpc++_alts.so.1.37.0-dev
lrwxrwxrwx  1       29 Mar  4 16:13 libgrpc++_error_details.so -> libgrpc++_error_details.so.15
-rwxrwxr-x  1     7416 Mar  4 16:13 libgrpc++_error_details.so.1.37.0-dev
lrwxrwxrwx  1       37 Mar  4 16:13 libgrpc++_error_details.so.15 -> libgrpc++_error_details.so.1.37.0-dev
lrwxrwxrwx  1       26 Mar  4 16:13 libgrpc++_reflection.so -> libgrpc++_reflection.so.15
-rwxrwxr-x  1  8104032 Mar  4 16:13 libgrpc++_reflection.so.1.37.0-dev
lrwxrwxrwx  1       34 Mar  4 16:13 libgrpc++_reflection.so.15 -> libgrpc++_reflection.so.1.37.0-dev
lrwxrwxrwx  1       24 Mar  4 16:13 libgrpc++_unsecure.so -> libgrpc++_unsecure.so.15
-rwxrwxr-x  1  2418208 Mar  4 16:13 libgrpc++_unsecure.so.1.37.0-dev
lrwxrwxrwx  1       32 Mar  4 16:13 libgrpc++_unsecure.so.15 -> libgrpc++_unsecure.so.1.37.0-dev
lrwxrwxrwx  1       13 Mar  4 16:13 libgrpc.so -> libgrpc.so.15
lrwxrwxrwx  1       17 Mar  4 16:13 libgrpc.so.15 -> libgrpc.so.15.0.0
-rwxrwxr-x  1 15887920 Mar  4 16:13 libgrpc.so.15.0.0
lrwxrwxrwx  1       24 Mar  4 16:13 libgrpc_csharp_ext.so -> libgrpc_csharp_ext.so.15
lrwxrwxrwx  1       28 Mar  4 16:13 libgrpc_csharp_ext.so.15 -> libgrpc_csharp_ext.so.15.0.0
-rwxrwxr-x  1    52272 Mar  4 16:13 libgrpc_csharp_ext.so.15.0.0
lrwxrwxrwx  1       28 Mar  4 16:12 libgrpc_plugin_support.so -> libgrpc_plugin_support.so.15
-rwxrwxr-x  1  8609840 Mar  4 16:12 libgrpc_plugin_support.so.1.37.0-dev
lrwxrwxrwx  1       36 Mar  4 16:12 libgrpc_plugin_support.so.15 -> libgrpc_plugin_support.so.1.37.0-dev
lrwxrwxrwx  1       22 Mar  4 16:13 libgrpc_unsecure.so -> libgrpc_unsecure.so.15
lrwxrwxrwx  1       26 Mar  4 16:13 libgrpc_unsecure.so.15 -> libgrpc_unsecure.so.15.0.0
-rwxrwxr-x  1  8170872 Mar  4 16:13 libgrpc_unsecure.so.15.0.0
lrwxrwxrwx  1       24 Mar  4 16:13 libgrpcpp_channelz.so -> libgrpcpp_channelz.so.15
-rwxrwxr-x  1 10808728 Mar  4 16:13 libgrpcpp_channelz.so.1.37.0-dev
lrwxrwxrwx  1       32 Mar  4 16:13 libgrpcpp_channelz.so.15 -> libgrpcpp_channelz.so.1.37.0-dev
lrwxrwxrwx  1       12 Mar  4 16:12 libupb.so -> libupb.so.15
lrwxrwxrwx  1       16 Mar  4 16:12 libupb.so.15 -> libupb.so.15.0.0
-rwxrwxr-x  1   228936 Mar  4 16:12 libupb.so.15.0.0
$ objdump -p libgrpc++.so.1.37.0-dev | grep SONAME
  SONAME               libgrpc++.so.15
$ readelf -a libgrpc++_reflection.so | grep NEED
  [ 6] .gnu.version_r    VERNEED          000000000023ba58  0023ba58
 0x0000000000000001 (NEEDED)             Shared library: [libpthread.so.0]
 0x0000000000000001 (NEEDED)             Shared library: [libgrpc++.so.15]
 0x0000000000000001 (NEEDED)             Shared library: [libstdc++.so.6]
 0x0000000000000001 (NEEDED)             Shared library: [libgcc_s.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libc.so.6]
 0x0000000000000001 (NEEDED)             Shared library: [ld-linux-x86-64.so.2]
 0x000000006ffffffe (VERNEED)            0x23ba58
 0x000000006fffffff (VERNEEDNUM)         5
```

As one can see, the GRPC artifacts produce and consume the version specific SONAME. Thus, it will be possible to use different runtime versions on the same root filesystem without a conflict.

As linked above, there was a lot of issues and confusion around the actual SONAME and the manually linking to `libgrpc++.so.1` in the past. But now that the right infrastructure is there it should be possible to fix this a right way.

Thanks

@markdroth

### CODE REVIEW

File: templates/CMakeLists.txt.template
Comment (jtattermusch) [inline templates/CMakeLists.txt.template:148]: the SO version of grpc++ is actually supposed to be 1.x.  the grpc library (=C core) is versioned differently (and is currently at version 15).  We'd have to have a really strong reason to change that since the change you're proposing is intuitively incorrect.

Comment (weltling) [reply templates/CMakeLists.txt.template]: Good point. The intention was just to make a differentiation on the SONAME. I've adapted as you suggest uggest to set `"${settings.cpp_version.major}.${settings.cpp_version.major}"`. It would look as below:

```
lrwxrwxrwx  1       17 Mar  5 10:39 libgrpc++.so -> libgrpc++.so.1.37
lrwxrwxrwx  1       23 Mar  5 10:39 libgrpc++.so.1.37 -> libgrpc++.so.1.37.0-dev
-rwxrwxr-x  1  2789096 Mar  5 10:39 libgrpc++.so.1.37.0-dev
lrwxrwxrwx  1       22 Mar  5 10:40 libgrpc++_alts.so -> libgrpc++_alts.so.1.37
lrwxrwxrwx  1       28 Mar  5 10:40 libgrpc++_alts.so.1.37 -> libgrpc++_alts.so.1.37.0-dev
-rwxrwxr-x  1   144864 Mar  5 10:40 libgrpc++_alts.so.1.37.0-dev
lrwxrwxrwx  1       31 Mar  5 10:39 libgrpc++_error_details.so -> libgrpc++_error_details.so.1.37
lrwxrwxrwx  1       37 Mar  5 10:39 libgrpc++_error_details.so.1.37 -> libgrpc++_error_details.so.1.37.0-dev
-rwxrwxr-x  1     7416 Mar  5 10:39 libgrpc++_error_details.so.1.37.0-dev
lrwxrwxrwx  1       28 Mar  5 10:40 libgrpc++_reflection.so -> libgrpc++_reflection.so.1.37
lrwxrwxrwx  1       34 Mar  5 10:40 libgrpc++_reflection.so.1.37 -> libgrpc++_reflection.so.1.37.0-dev
-rwxrwxr-x  1  1120672 Mar  5 10:40 libgrpc++_reflection.so.1.37.0-dev
lrwxrwxrwx  1       26 Mar  5 10:39 libgrpc++_unsecure.so -> libgrpc++_unsecure.so.1.37
lrwxrwxrwx  1       32 Mar  5 10:39 libgrpc++_unsecure.so.1.37 -> libgrpc++_unsecure.so.1.37.0-dev
-rwxrwxr-x  1  2418208 Mar  5 10:39 libgrpc++_unsecure.so.1.37.0-dev
```

Thanks

Comment (weltling) [reply templates/CMakeLists.txt.template]: And same way, checking the artifacts:

```bash
$ objdump -p libgrpc++.so | grep SONAME
  SONAME               libgrpc++.so.1.37
$ readelf -a libgrpc++_reflection.so | grep NEED
  [ 6] .gnu.version_r    VERNEED          0000000000058200  00058200
 0x0000000000000001 (NEEDED)             Shared library: [libgrpc++.so.1.37]
 0x0000000000000001 (NEEDED)             Shared library: [libprotobuf.so.3.15.2.0]
 0x0000000000000001 (NEEDED)             Shared library: [libstdc++.so.6]
 0x0000000000000001 (NEEDED)             Shared library: [libgcc_s.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libc.so.6]
 0x000000006ffffffe (VERNEED)            0x58200
 0x000000006fffffff (VERNEEDNUM)         3
```

Thanks

Comment (weltling) [reply templates/CMakeLists.txt.template]: Rebased to latest master. @jtattermusch please let me know if there's a better approach to address this issue. Perhaps putting the whole version would be even better, fe see what boost does. A proper handling of SONAME would be really useful for downstreams and similar situations.

Thanks

**Comment by nicolasnoble** [comment]
I don't mind moving forward with this one, but be warned that while we're doing best efforts to be aware of ABI breakages, we offer no real guarantees, and so this may be doing more harm than good by implying we're totally on top of ABI, which we are not.

**Review by jtattermusch — APPROVED** [review]
I restarted the tests. 
This new approach looks fine - LGTM as long as the tests pass. 

I'd like @nicolasnoble  to approve as well before we merge this.

**Comment by nicolasnoble** [comment]
I think I'd want to see a warning somewhere in a readme that we're only doing best effort to bump soname revision numbers during abi breakages, and that we can't make hard guarantees that this effectively reflects any sort of ABI stability across the same SONAME version. Otherwise, yes, LGTM.

@drfloob may have had an idea for this?

**Comment by weltling** [comment]
I've added a note to `CONCEPTS.md` explaining the matter. That file looked as closest. Please let me know if there's a better place.

Thanks

**Comment by drfloob** [comment]
Thanks, Anatol! I had planned to write up something on ABI stability, looks good. It seems maybe unfriendly of us to claim we're making ABI incompatible changes at least every 6 weeks on our release cadence, more often if we catch an actual ABI breakage otherwise. But it does solve the problem of not knowing (i.e. not explicitly testing for) when we're breaking the ABI.

One change I'd suggest: `CONCEPTS.md` is more about what gRPC does in the abstract. I think a better place for that explanation might be in `BUILDING.md`, where we already have related messaging:

> Using gRPC C++ as a DLL is not recommended, but you can still enable it by running `cmake` with `-DBUILD_SHARED_LIBS=ON`.

File: CONCEPTS.md
Comment (jtattermusch) [inline CONCEPTS.md:40]: this definitely doesn't belong here.

I think you can add a note to BUILDING.md (if you find no good place for it, feel free to append a section at the end of it).

Also, I think think it would be better to put this more in context. 1.) The SONAME only affects our cmake builds and 2.) the remark you wrote only describes the relationship between ABI compatibility and the SONAME.

But somehow as-is, this section sounds as if you're making general claims about ABI compatibility of gRPC and gRPC++ and that's not the intent. Please rephrase to make it clear what's the scope of the information this provides (e.g. we could call it "A note on SONAME and its ABI compatibility implications in the cmake build".

Comment (weltling) [reply CONCEPTS.md]: Yep, moved as suggested and changed the title. I used the phasing almost as in https://github.com/grpc/grpc/pull/25617#issuecomment-799757679, whereby OFC it's only about this specific situation. As AJ told previously to have an intention writing up something with more details, the current version should be just to document the change as requested.

Thanks.

Comment (jtattermusch) [reply CONCEPTS.md]: Thank for updating the docs.

**Review by jtattermusch — APPROVED** [review]
I took another look and LGTM. 

@nicolasnoble  @drfloob  care to rubberstamp this as well?

**Review by drfloob — APPROVED** [review]
LGTM. Thanks for the improvement.

**Comment by weltling** [comment]
Thanks a lot for accepting this, great collaboration!

**Pull request merged**
